aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,433 @@
+/// This defines a proposed Aptos object model with the intent to provide the following properties:
+/// * Decouple owner from data
+///   * Objects are stored as resources within an account
+///   * Ownership is defined in the object itself, no moving around stores
+/// * Heterogeneous types stored in a single container
+///   * Decoupled data from identifier allows for referencing ambiguous types
+///   * Ambiguous types can be homogeneous thus bypassing restrictions on heterogeneous collections
+/// * Make data immobile, while leaving ownership flexible
+///   * Resources cannot be easily moved without explicit permission by the module that defines them
+///   * The various refs enable extensible definition on ownership
+/// * Data is globally accessible
+///   * Data resides in resources which cannot be moved and therefore are always accessible
+///   * In contrast, storable struct can be placed into an inaccessible space
+/// * Objects can be the source of events
+///   * When data is mutated in a struct, ideally that struct would emit an event.
+///   * The object model allows for events to be directly coupled with data.
+/// * Objects definition / structure can prevented after creation
+///   * At creation, the creator has the ability to configure the account as desired
+///   * Afterward, the creator can delegate signer ability to continue to change definition
+///     but it can also be discarded or access prevented to ensure that an object maintains
+///     consistent shape.
+/// * Secure and guaranteed to not collide with existing accounts.
+///   * Each address is generated in a collision-resistant fashion guaranteeing legitimate
+///     accounts cannot be created within the same address.
+///   * Two objects cannot be co-located due to an address being able to only store a single Object
+///
+/// TODO:
+/// * There is no means to borrow an object or a reference to an object. We are exploring how to
+///   make it so that a reference to a global object can be returned from a function.
+module aptos_framework::object {
+    use std::bcs;
+    use std::error;
+    use std::hash;
+    use std::signer;
+    use std::vector;
+
+    use aptos_framework::account;
+    use aptos_framework::event;
+    use aptos_framework::from_bcs;
+    use aptos_framework::guid;
+
+    /// An object already exists at this address
+    const EOBJECT_EXISTS: u64 = 0;
```

Comment:
nit: Can we not avoid using 0? 0 is treated in many serialization formats such as in. protobuf 3 as being the default value for enum if absent.